auth/developer-single-sign-on-across-mobile-services.md

```diff
@@ -0,0 +1,59 @@
+
+## Problem Description
+
+As a developer builds up their Project by provisioning different services, they will likely need to setup or configure ascpects of those services. For example, if a developer provisions the Push Notifications service, they will need to configure Push credentials for Mobile Apps. Some services may also provide additional info or dasboards. Accessing these dashboards & performing setup or configuration will likely require signing into those services via some auth mechanism.
+
+Each service will have 1 or more of the following:
+
+* it's own auth mechanism
+* ability to configure an external auth mechanism
+* no auth mechanism
+
+This could cause an inconsistent and poor user experience for the following reasons:
+
+* different auth flow for each service
+* different auth credentials for each service, managed independently
+
+Additionaly, a service that has no auth mechanism, e.g. Prometheus, will need to be secured in some manner if exposed externally.
+
+## Proposed Solution
+
+### OpenShift as the Single Sign-On Provider
+
+All Services will be running on OpenShift, or 'ordered' from the OpenShift Catalog. OpenShift is the only constant between the Developer and the Services. The Developer will need to be signed in to OpenShift. Leveraging this fact and having OpenShift be the single auth provider for all Services will address the main points in the problem description.
+
+The flow for signing in to each Service will be the same i.e. an OAuth flow where the user signs in & authorises the Service with OpenShift. No additional credentials will be required.
+
+For cases where no auth mechanism is provided with a service, a reverse proxy approach that uses OpenShift OAuth, as shown in Option 3 below, can be used.
+
+### Provider Integration Options
+
+Choosing how to integrate a Service with OpenShift OAuth will depend on what the Service is capable of already, and how extensible it is. The below options show practical examples of how this integration was done. There will likely be tweaks and other options required depending on the service. These options serve as references for how it could be done.
+
+#### Option 1: Configure OAuth in the Service
+
+If a Service has the ability to an OAuth provider, use it. For example, in Keycloak it is possible to use 'Identity Brokering' to defer to OpenShift Auth for who can administer that Keycloak instance. See this [Keycloak Identity Brokering Blog Post](https://developers.redhat.com/blog/2017/12/06/keycloak-identity-brokering-openshift/) for details.
+
+However, with Keycloak this method is limited to only automatically setting up the initial permissions of *all* users who can successfully authenticate. There is no permissions mapping between OpenShift roles & Keycloak administration roles. Additionally, *any* OpenShift user will be able to sign in to Keycloak. Manual configuration is required to only allow certain users access and limit their roles.
```

Comment:
is the manual configuration on a per user level or is it something that we could look to do from the APB?

Reply:
Per user level.
Working with @wei-lee to figure out if this could be automated. Will be chatting with keycloak dev members too